lib/diamond/utils.ts

```diff
@@ -0,0 +1,305 @@
+export function getFacets(contracts: Contract[]): Facet[] {
+  return contracts.map((contract) => {
+    return {
+      target: contract.address,
+      selectors: getSelectors(contract),
+    };
+  });
+}
+
+// returns true if the selector is found in the facets
+export function selectorExistsInFacets(
+  selector: string,
+  facets: Facet[],
+): boolean {
+  for (const facet of facets) {
+    if (facet.selectors.includes(selector)) return true;
+  }
+  return false;
+}
+
+// preview FacetCut which adds unregistered selectors
+export async function addUnregisteredSelectors(
+  diamond: IDiamondReadable,
+  contracts: Contract[],
+  only: FacetFilter[] = [],
+  except: FacetFilter[] = [],
+): Promise<FacetCut[]> {
+  validateFilters(only, except);
+
+  const diamondFacets: Facet[] = await diamond.facets();
+  const facets = getFacets(contracts);
+
+  let selectorsAdded = false;
+  let facetCuts: FacetCut[] = [];
+
+  // if facet selector is unregistered then it should be added to the diamond.
+  for (const facet of facets) {
+    for (const selector of facet.selectors) {
+      const target = facet.target;
+
+      if (
+        target !== diamond.address &&
+        selector.length > 0 &&
+        !selectorExistsInFacets(selector, diamondFacets) &&
+        selectorIsFiltered(only, except, target, selector)
+      ) {
+        facetCuts.push(
+          printFacetCuts(facet.target, [selector], FacetCutAction.ADD),
+        );
+
+        selectorsAdded = true;
+      }
+    }
+  }
+
+  if (!selectorsAdded) {
+    throw new Error('No selectors were added to FacetCut');
+  }
+
+  return groupFacetCuts(facetCuts);
+}
+
+// preview FacetCut which replaces registered selectors with unregistered selectors
+export async function replaceRegisteredSelectors(
+  diamond: IDiamondReadable,
+  contracts: Contract[],
+  only: FacetFilter[] = [],
+  except: FacetFilter[] = [],
+): Promise<FacetCut[]> {
+  validateFilters(only, except);
+
+  const diamondFacets: Facet[] = await diamond.facets();
+  const facets = getFacets(contracts);
+
+  let selectorsReplaced = false;
+  let facetCuts: FacetCut[] = [];
+
+  // if a facet selector is registered with a different target address, the target will
+  // be replaced
+  for (const facet of facets) {
+    for (const selector of facet.selectors) {
+      const target = facet.target;
+      const oldTarget = await diamond.facetAddress(selector);
+
+      if (
+        target != oldTarget &&
+        target != AddressZero &&
+        target != diamond.address &&
+        selector.length > 0 &&
+        selectorExistsInFacets(selector, diamondFacets) &&
+        selectorIsFiltered(only, except, target, selector)
+      ) {
+        facetCuts.push(
+          printFacetCuts(target, [selector], FacetCutAction.REPLACE),
+        );
+
+        selectorsReplaced = true;
+      }
+    }
+  }
+
+  if (!selectorsReplaced) {
+    throw new Error('No selectors were replaced in FacetCut');
```

Comment:
The errors are thrown if _all_ changes are skipped, but I think we'll want to throw if _any_ change is skipped.  Need to think about this though.

Reply:
Would this be the result of using filters or something else?